panels/TimelinePanel.php

```diff
@@ -0,0 +1,237 @@
+            'panel' => $this,
+            'dataProvider' => $dataProvider,
+            'searchModel' => $searchModel,
+        ]);
+    }
+
+    /**
+     * Returns an array of models that represents logs of the current request.
+     * Can be used with data providers, such as \yii\data\ArrayDataProvider.
+     *
+     * @param boolean $refresh if need to build models from log messages and refresh them.
+     * @return array models
+     */
+    protected function getModels($refresh = false)
+    {
+        if ($this->_models === null || $refresh) {
+            $this->_models = [];
+            $logMessages = isset($this->module->panels['log']->data['messages']) ? $this->module->panels['log']->data['messages'] : [];
+            $profilingMessages = isset($this->module->panels['profiling']->data['messages']) ? $this->module->panels['profiling']->data['messages'] : [];
+
+            foreach ($logMessages as $key => $message) {
+                if (
+                    ($message[1] === Logger::LEVEL_PROFILE_BEGIN || $message[1] === Logger::LEVEL_PROFILE_END)
+                    || ($message[1] === Logger::LEVEL_INFO && $message[2] == 'yii\db\Command::query')
+                ) {
+                    continue;
+                }
+                $this->_models[] = [
+                    'info' => $message[0],
+                    'level' => $message[1],
+                    'category' => $message[2],
+                    'timestamp' => ($message[3] * 1000), // time in milliseconds
+                    'trace' => $message[4],
+                    'duration' => null
+                ];
+            }
+
+            foreach (Yii::getLogger()->calculateTimings($profilingMessages) as $profile) {
+                $profile['timestamp'] = $profile['timestamp'] * 1000;
+                $profile['duration'] = $profile['duration'] * 1000;
+                $this->_models[] = $profile;
+            }
+        }
+        return $this->_models;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function save()
+    {
+        return [
+            'start' => YII_BEGIN_TIME,
+            'end' => microtime(true),
+        ];
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function load($data)
+    {
+        parent::load($data);
+        $this->initTimestamp();
+    }
+
+    /**
+     * set timestamps
+     */
+    protected function initTimestamp()
+    {
+        $this->timestamps = [
+            $this->data['start'] * 1000,
+            $this->data['end'] * 1000,
+        ];
+        if (isset($this->module->panels['profiling']->data['time'])) {
+            $this->timestamps[2] = $this->module->panels['profiling']->data['time'] * 1000;
+        } else {
+            $this->timestamps[2] = $this->timestamps[1] - $this->timestamps[0];
+        }
+    }
+
+    /**
+     * @param int $line
+     * @return array
+     */
+    public function getRulers($line = 10)
+    {
+        --$line;
+        $data[0] = 0;
```

Comment:
That's rather cryptic. If you want to initialize `$data`, do it like the following:

```php
$data = [0];
```